### PR TITLE
Add masking extension

### PR DESCRIPTION
**Pull request by arthuro555**
Add an extension for using PIXI sprites masking.

### CODE REVIEW

**Comment by arthuro555** [comment]
Ready for review @4ian

**Comment by Bouh** [comment]
Interesting extension and easy to use!

You use an object for mask a part of another object. I understand why you do this, you have not other choice.
If in the function parameter we can have a ressource selector 🙏 , you will be able to get the mask image directlly from the ressource manager. This have more sense than select an object for mask another object.

**Comment by arthuro555** [comment]
> Version "definitive", we need always to use a version number, PIXI can change how mask work and if an user use an old version of your extension on a recent build, this one will be broken, the user cannot see the difference between your versions because there is no number version.

Yes I changed my mind about the version but was too lazy to correct it before your comment :/


> If in the function parameter we can have a ressource selector 🙏 , you will be able to get the mask image directlly from the ressource manager. This have more sense than select an object for mask another object. (With a resource selector i imagine already the button extension!)

Sadly, masking requires a sprite in PIXI on both sides.

**Comment by Bouh** [comment]
I was pretty sure to just need Pixi texture.

File: Extensions/MaskingExtension.json [hunk 1]
```diff
@@ -0,0 +1,113 @@
+{
+  "author": "Arthur Pacaud (arthuro555)",
+  "description": "When masked, the masked sprite is only visible semitransparently through the mask.",
+  "extensionNamespace": "",
+  "fullName": "Sprite Masking",
+  "name": "Masking",
+  "shortDescription": "Permits masking of sprites",
+  "tags": "masking, javascript, PIXI, sprites",
+  "version": "definitive",
```
Comment (4ian) [inline Extensions/MaskingExtension.json:9]: Version should always follow https://semver.org/, so name it 1.0.0

File: Extensions/MaskingExtension.json [hunk 2]
```diff
@@ -0,0 +1,113 @@
+{
+  "author": "Arthur Pacaud (arthuro555)",
+  "description": "When masked, the masked sprite is only visible semitransparently through the mask.",
+  "extensionNamespace": "",
+  "fullName": "Sprite Masking",
+  "name": "Masking",
+  "shortDescription": "Permits masking of sprites",
```
Comment (4ian) [inline Extensions/MaskingExtension.json:7]: Allow to use a sprite object to mask another.

File: Extensions/MaskingExtension.json [hunk 3]
```diff
@@ -0,0 +1,113 @@
+{
+  "author": "Arthur Pacaud (arthuro555)",
+  "description": "When masked, the masked sprite is only visible semitransparently through the mask.",
+  "extensionNamespace": "",
+  "fullName": "Sprite Masking",
+  "name": "Masking",
```
Comment (4ian) [inline Extensions/MaskingExtension.json:6]: Let's name the extension `SpriteMasking`.

File: Extensions/MaskingExtension.json [hunk 4]
```diff
@@ -0,0 +1,113 @@
+{
```
Comment (4ian) [inline Extensions/MaskingExtension.json:1]: The name of the file must be the name of the extension. I suggest `SpriteMasking.json`

File: Extensions/MaskingExtension.json [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+          "type": "objectList"
+        }
+      ],
+      "objectGroups": []
+    },
+    {
+      "description": "Check for masking on a sprite",
+      "fullName": "Is Masked?",
+      "functionType": "Condition",
+      "name": "Masked",
```
Comment (4ian) [inline Extensions/MaskingExtension.json:85]: Let's name this `IsMasked` to emphasize this is a condition

File: Extensions/MaskingExtension.json [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+          "name": "Masked",
+          "optional": false,
+          "supplementaryInformation": "Sprite",
+          "type": "objectList"
+        }
+      ],
+      "objectGroups": []
+    },
+    {
+      "description": "Check for masking on a sprite",
```
Comment (4ian) [inline Extensions/MaskingExtension.json:82]: Check if a sprite is masked by another.

File: Extensions/MaskingExtension.json [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+          "name": "Mask",
+          "optional": false,
+          "supplementaryInformation": "Sprite",
+          "type": "objectList"
+        }
+      ],
+      "objectGroups": []
+    },
+    {
+      "description": "Cancels masking of a sprite by another",
```
Comment (4ian) [inline Extensions/MaskingExtension.json:52]: nitpicking: period at the end of the sentence.

File: Extensions/MaskingExtension.json [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  "description": "When masked, the masked sprite is only visible semitransparently through the mask.",
+  "extensionNamespace": "",
+  "fullName": "Sprite Masking",
+  "name": "Masking",
+  "shortDescription": "Permits masking of sprites",
+  "tags": "masking, javascript, PIXI, sprites",
+  "version": "definitive",
+  "eventsFunctions": [
+    {
+      "description": "Defines a sprite as mask of another",
```
Comment (4ian) [inline Extensions/MaskingExtension.json:12]: nitpicking: period at the end of the sentence.

**Comment by 4ian** [comment]
That's a very interesting extension, thanks for working on this! Finally got the time to review it.

> Yes I changed my mind about the version but was too lazy to correct it before your comment :/

Remember that you're the owner of your PRs, and that they cost time to review - I typically spend *hours* each day reviewing them - that I could invest in finishing features instead.
Getting the small details right is an important part of submitting a PR ;)

**Comment by arthuro555** [comment]
Sorry, I'll try to be more careful about every details.

**Comment by 4ian** [comment]
> Graphics (=> shape painter but you probably already know) are fully invisible but not sprites.

Alright then! Seems that it's using the red and alpha channel to do the masking according to the docs.
We can keep the description then.

If you remove the gdjs namespace then we can go ahead with this :)

**Comment by arthuro555** [comment]
> If you remove the gdjs namespace then we can go ahead with this :)

I already did 👀

**Comment by 4ian** [comment]
Proofread the changed files please :)

**Comment by arthuro555** [comment]
Oh yes forgot to say I corrected the header too sorry for missing that one.

**Comment by 4ian** [comment]
There was still a gdjs missing. Always proofread your PR like a reviewer would do :) Here a simple search for "gdjs" would have done the job - but I did miss it because I was assuming you would have done it.

**Comment by arthuro555** [comment]
I am sorry but also very confused. I reread the changes in the changes tab and saw it was in the header, so I thought regenerating the header with the node script would have fixed it.

**Pull request merged**